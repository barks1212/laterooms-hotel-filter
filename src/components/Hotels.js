import React from 'react';
import hotelList from '../hotel-list.json';
import SearchFilters from './Search-filters';
import Hotel from './Hotel';
import Sort from './Sort';

class Hotels extends React.Component {
  state = {
    hotelList,
    sort: 'ascending',
    filters: {
      bar: false,
      gym: false,
      spa: false,
      pool: false,
      restaurant: false,
      'car park': false,
      'full board': false,
      'half board': false
    }
  }
  render() {
    return (
      <section className="mainBody">
        <SearchFilters filters={this.state.filters} hotelFilter={this.hotelFilter}/>
        <section className="mainBodyHotels">
          <Sort sort={this.state.sort} sorter={this.sorter} hotelList={this.state.hotelList} />
          {this.state.hotelList.map((hotel, i) => {
            return (
              <section className="" key={i}>
                <Hotel hotel={hotel} />
              </section>
            );
          })}
        </section>
      </section>
    );
  }

  sorter = (order, hotels) => {
    this.setState({ sort: order });

    if (order === 'ascending') {
      let ascendingHotels = hotels.sort((a, b) => {
        return a.StarRating - b.StarRating;
      });
      this.setState({ hotelList: ascendingHotels });
    }

    if (order === 'descending') {
      let descendingHotels = hotels.sort((a, b) => {
        return b.StarRating - a.StarRating;
      });
      this.setState({ hotelList: descendingHotels });
    }
  }

  hotelFilter = (fac /*filters*/) => {
    let newFilters = Object.assign(this.state.filters, { [fac]: !this.state.filters[fac] });
    this.setState({ filters: newFilters });
    
    let checkedChecker = [];
    for (let key in this.state.filters) {
      if (this.state.filters[key]) checkedChecker.push(key);
    }

    let newHotelList = hotelList.filter((hotel) => {
      return checkedChecker.every((facility) => {
        return hotel.Facilities.includes(facility);
      });
    });
    return this.sorter(this.state.sort, newHotelList);
  }

}

export default Hotels;
