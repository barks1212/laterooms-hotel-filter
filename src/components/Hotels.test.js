import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Hotels from './Hotels';

describe('<Hotels>', () => {
  it('renders as expected without crashing and matches snapshot', () => {
    const tree = renderer.create(<Hotels />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('can sort hotels based on stars in ascending order', () => {
    const wrapper = mount(<Hotels />);
    const hotels = wrapper.state().hotelList;
    wrapper.find('button').at(1).simulate('click');
    expect(hotels[0].StarRating).toEqual(2);
    expect(hotels[hotels.length - 1].StarRating).toEqual(5);
  });

  it('can sort hotels based on stars in descending order', () => {
    const wrapper = mount(<Hotels />);
    const hotels = wrapper.state().hotelList;
    wrapper.find('button').at(0).simulate('click');
    expect(hotels[0].StarRating).toEqual(5);
    expect(hotels[hotels.length - 1].StarRating).toEqual(2);
  });

  it('filters hotels based on their facilities', () => {
    const wrapper = mount(<Hotels />);
    wrapper.find('input').at(0).simulate('change');
    const facility = wrapper.find('input').at(0).getElement().props.value;
    wrapper.state().hotelList.forEach((hotel) => {
      expect(hotel.Facilities.includes(facility)).toBe(true);
    });
    expect(wrapper.state().hotelList).toHaveLength(5);
    wrapper.find('input').at(1).simulate('change');
    expect(wrapper.state().hotelList).toHaveLength(3);
  });
});