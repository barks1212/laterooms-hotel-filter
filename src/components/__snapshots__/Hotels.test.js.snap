// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hotels> renders as expected without crashing and matches snapshot 1`] = `
<section
  className="mainBody"
>
  <section
    className="filters"
  >
    <section
      className="columns is-mobile is-gapless"
    >
      <section
        className="column is-one-half"
        id="columnOne"
      >
        <label
          className="checkbox"
          id="checkboxLeft"
        >
          <input
            type="checkbox"
            value="bar"
          />
          <span
            className="is-size-7"
          >
            Bar
          </span>
        </label>
        <label
          className="checkbox"
          id="checkboxLeft"
        >
          <input
            type="checkbox"
            value="gym"
          />
          <span
            className="is-size-7"
          >
            Gym
          </span>
        </label>
        <label
          className="checkbox"
          id="checkboxLeft"
        >
          <input
            type="checkbox"
            value="spa"
          />
          <span
            className="is-size-7"
          >
            Spa
          </span>
        </label>
        <label
          className="checkbox"
          id="checkboxLeft"
        >
          <input
            type="checkbox"
            value="pool"
          />
          <span
            className="is-size-7"
          >
            Pool
          </span>
        </label>
      </section>
      <section
        className="column is-one-half"
        id="columnTwo"
      >
        <label
          className="checkbox"
          id="checkboxRight"
        >
          <input
            type="checkbox"
            value="restaurant"
          />
          <span
            className="is-size-7"
          >
            Restaurant
          </span>
        </label>
        <label
          className="checkbox"
          id="checkboxRight"
        >
          <input
            type="checkbox"
            value="car park"
          />
          <span
            className="is-size-7"
          >
            Car park
          </span>
        </label>
        <label
          className="checkbox"
          id="checkboxRight"
        >
          <input
            type="checkbox"
            value="full board"
          />
          <span
            className="is-size-7"
          >
            Full board
          </span>
        </label>
        <label
          className="checkbox"
          id="checkboxRight"
        >
          <input
            type="checkbox"
            value="half board"
          />
          <span
            className="is-size-7"
          >
            Half board
          </span>
        </label>
      </section>
    </section>
  </section>
  <section
    className="mainBodyHotels"
  >
    <section
      className="sort is-paddingless is-marginless"
    >
      <span
        className="is-paddingless has-text-weight-bold"
      >
         Sort by 
      </span>
      <button
        className="button is-small is-paddingless is-marginless is-text"
        id="sortButton"
      >
        Descending
      </button>
      |
      <button
        className="button is-small is-paddingless is-marginless is-text"
        id="sortButton"
      >
        Ascending
      </button>
    </section>
    <section
      className=""
    >
      <section
        className="hotels"
      >
        <section
          className="card is-mobile"
          id="hotelCard"
        >
          <figure
            className="image is-128x128"
          >
            <img
              alt=""
              id="hotelPic"
              src="https://s-ec.bstatic.com/images/hotel/max1024x768/308/30801707.jpg"
            />
          </figure>
          <section
            className="nameAndRating"
          >
            <h2
              className="subtitle has-text-weight-bold is-size-7 is-marginless"
            >
              Yacht Harbour
            </h2>
            <span
              className="starRating"
            >
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
            </span>
          </section>
        </section>
        <section
          className="hotelContent"
        >
          <section
            className="container is-mobile"
            id="facilities"
          >
            <p
              className="facility is-size-6 is-mobile"
            >
              Car park
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Spa
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Pool
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Gym
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Full board
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Bar
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Restaurant
            </p>
          </section>
        </section>
      </section>
    </section>
    <section
      className=""
    >
      <section
        className="hotels"
      >
        <section
          className="card is-mobile"
          id="hotelCard"
        >
          <figure
            className="image is-128x128"
          >
            <img
              alt=""
              id="hotelPic"
              src="https://t-ec.bstatic.com/images/hotel/max1024x768/259/25947064.jpg"
            />
          </figure>
          <section
            className="nameAndRating"
          >
            <h2
              className="subtitle has-text-weight-bold is-size-7 is-marginless"
            >
              The Adriana
            </h2>
            <span
              className="starRating"
            >
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
            </span>
          </section>
        </section>
        <section
          className="hotelContent"
        >
          <section
            className="container is-mobile"
            id="facilities"
          >
            <p
              className="facility is-size-6 is-mobile"
            >
              Car park
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Pool
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Bar
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Half board
            </p>
          </section>
        </section>
      </section>
    </section>
    <section
      className=""
    >
      <section
        className="hotels"
      >
        <section
          className="card is-mobile"
          id="hotelCard"
        >
          <figure
            className="image is-128x128"
          >
            <img
              alt=""
              id="hotelPic"
              src="https://t-ec.bstatic.com/xdata/images/hotel/square200/20927575.jpg?k=682cb00a8765e55512f0308073de3949fd0c1838601455fc5fac40b353ccb104&o="
            />
          </figure>
          <section
            className="nameAndRating"
          >
            <h2
              className="subtitle has-text-weight-bold is-size-7 is-marginless"
            >
              Hotel Fortuna
            </h2>
            <span
              className="starRating"
            >
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
            </span>
          </section>
        </section>
        <section
          className="hotelContent"
        >
          <section
            className="container is-mobile"
            id="facilities"
          >
            <p
              className="facility is-size-6 is-mobile"
            >
              Car park
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Half board
            </p>
          </section>
        </section>
      </section>
    </section>
    <section
      className=""
    >
      <section
        className="hotels"
      >
        <section
          className="card is-mobile"
          id="hotelCard"
        >
          <figure
            className="image is-128x128"
          >
            <img
              alt=""
              id="hotelPic"
              src="https://s-ec.bstatic.com/xdata/images/hotel/square200/103265799.jpg?k=bc8c01f4f5ce8911a10c788cb4864790756546ab0cc0a81c10c262f956efd84d&o="
            />
          </figure>
          <section
            className="nameAndRating"
          >
            <h2
              className="subtitle has-text-weight-bold is-size-7 is-marginless"
            >
              Aparthotel Pharia
            </h2>
            <span
              className="starRating"
            >
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
            </span>
          </section>
        </section>
        <section
          className="hotelContent"
        >
          <section
            className="container is-mobile"
            id="facilities"
          >
            <p
              className="facility is-size-6 is-mobile"
            >
              Full board
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Restaurant
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Car park
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Gym
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Pool
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Bar
            </p>
          </section>
        </section>
      </section>
    </section>
    <section
      className=""
    >
      <section
        className="hotels"
      >
        <section
          className="card is-mobile"
          id="hotelCard"
        >
          <figure
            className="image is-128x128"
          >
            <img
              alt=""
              id="hotelPic"
              src="https://s-ec.bstatic.com/xdata/images/hotel/square200/41476685.jpg?k=d873a52b1d820f8e2f991bb015dc2ec9f64e6070e4890a117d2533b20fcd0e95&o="
            />
          </figure>
          <section
            className="nameAndRating"
          >
            <h2
              className="subtitle has-text-weight-bold is-size-7 is-marginless"
            >
              Amfora Grand Beach
            </h2>
            <span
              className="starRating"
            >
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
            </span>
          </section>
        </section>
        <section
          className="hotelContent"
        >
          <section
            className="container is-mobile"
            id="facilities"
          >
            <p
              className="facility is-size-6 is-mobile"
            >
              Car park
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Half board
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Gym
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Pool
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Spa
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Bar
            </p>
          </section>
        </section>
      </section>
    </section>
    <section
      className=""
    >
      <section
        className="hotels"
      >
        <section
          className="card is-mobile"
          id="hotelCard"
        >
          <figure
            className="image is-128x128"
          >
            <img
              alt=""
              id="hotelPic"
              src="https://t-ec.bstatic.com/xdata/images/hotel/square200/75624593.jpg?k=4ffeb58898091a00f7f5c63852905e4e429c0d6c63a05da3a2f080f6c5656204&o="
            />
          </figure>
          <section
            className="nameAndRating"
          >
            <h2
              className="subtitle has-text-weight-bold is-size-7 is-marginless"
            >
              Hotel Podstine
            </h2>
            <span
              className="starRating"
            >
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
            </span>
          </section>
        </section>
        <section
          className="hotelContent"
        >
          <section
            className="container is-mobile"
            id="facilities"
          >
            <p
              className="facility is-size-6 is-mobile"
            >
              Full board
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Bar
            </p>
            <p
              className="facility is-size-6 is-mobile"
            >
              Restaurant
            </p>
          </section>
        </section>
      </section>
    </section>
    <section
      className=""
    >
      <section
        className="hotels"
      >
        <section
          className="card is-mobile"
          id="hotelCard"
        >
          <figure
            className="image is-128x128"
          >
            <img
              alt=""
              id="hotelPic"
              src="https://s-ec.bstatic.com/xdata/images/hotel/square200/58465884.jpg?k=4ccb617b3b004b9111f9657b7d711f7b32277a29aecb0d17496a7dfcbf8ae2c3&o="
            />
          </figure>
          <section
            className="nameAndRating"
          >
            <h2
              className="subtitle has-text-weight-bold is-size-7 is-marginless"
            >
              Delfin
            </h2>
            <span
              className="starRating"
            >
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
              <i
                aria-hidden="true"
                className="fa fa-star"
                id="stars"
              />
            </span>
          </section>
        </section>
        <section
          className="hotelContent"
        >
          <section
            className="container is-mobile"
            id="facilities"
          >
            <p
              className="facility is-size-6 is-mobile"
            >
              Car park
            </p>
          </section>
        </section>
      </section>
    </section>
  </section>
</section>
`;
